core/src/main/scala/kafka/server/DynamicBrokerConfig.scala

```diff
@@ -449,3 +456,81 @@ class DynamicLogConfig(logManager: LogManager) extends Reconfigurable with Loggi
     }
   }
 }
+
+class DynamicMetricsReporters(brokerId: Int, dynamicConfig: DynamicBrokerConfig) extends Reconfigurable {
+
+  private val propsOverride = Map[String, AnyRef](KafkaConfig.BrokerIdProp -> brokerId.toString)
+  private val currentReporters = mutable.Map[String, MetricsReporter]()
+  private var metrics: Metrics = _
+  initialize()
+
+  private def initialize() {
+    dynamicConfig.addReconfigurable(this)
+    val reporters = dynamicConfig.currentKafkaConfig.getList(KafkaConfig.MetricReporterClassesProp)
+    createReporters(reporters, Collections.emptyMap[String, Object], None)
+  }
+
+  private[server] def initMetrics(metrics: Metrics): Unit = {
+    this.metrics = metrics
+  }
+
+  private[server] def currentMetricsReporters: List[MetricsReporter] = currentReporters.values.toList
+
+  override def configure(configs: util.Map[String, _]): Unit = {}
+
+  override def reconfigurableConfigs(): util.Set[String] = {
+    val configs = new util.HashSet[String]()
+    configs.add(KafkaConfig.MetricReporterClassesProp)
+    currentReporters.values.foreach {
+      case reporter: Reconfigurable => configs.addAll(reporter.reconfigurableConfigs)
+      case _ =>
+    }
+    configs
+  }
+
+  override def validateReconfiguration(configs: util.Map[String, _]): Boolean = {
```

Comment:
It's a little confusing that validation failure is signaled both by the return value and an exception. Do we need both?

Reply:
As an external API, I thought a boolean is better. For internal validation, I am using exceptions where the message gets propagated to clients (that doesn't feel very safe to do for custom code where exceptions may potentially contain sensitive information like passwords). I have modified `BrokerReconfigurable` to only throw exceptions.

Reply:
Yes, I think that is fine. Of course plugins may want to propagate a message as well. I guess we could do something like we do with the policies and let the plugin raise `ConfigValidationException` (or something like that).

Reply:
Internally, `ConfigException` is used, which was being translated to `InvalidRequestException` for other config updates (like topic configs). So plugins could throw that exception. I don't mind removing the return value from `validateReconfiguration` if it is better to document the exception instead. Let me know what you think - perhaps I could make that update in the next PR.

Reply:
Yeah, we can consider separately. I think I slightly prefer documenting the use of `ConfigException` for consistency, but I don't feel too strongly about it.